c-dependencies/js-compute-runtime/js-compute-builtins.cpp

```diff
@@ -3625,6 +3630,388 @@ bool response_started(JSObject *self) {
+  }
+
+  // 8.3 If buffer has accumulated 24 bits, interpret them as three 8-bit big-endian numbers. Append
+  // three bytes with values equal to those numbers to output, in the same order, and then empty
+  // buffer.
+  output += (uint8_t(w << 2 | x >> 4));
+  output += (uint8_t(x << 4 | y >> 2));
+  output += (uint8_t(y << 6 | z));
+  return mozilla::Ok();
+}
+
+inline JS::Result<mozilla::Ok> base64Decode3to2(std::string_view input, std::string &output) {
+  uint8_t w, x, y;
+  // 8.1 Find the code point pointed to by position in the second column of Table 1: The Base 64
+  // Alphabet of RFC 4648. Let n be the number given in the first cell of the same row. [RFC4648]
+  if (!base64CharacterToValue(input[0], &w) || !base64CharacterToValue(input[1], &x) ||
+      !base64CharacterToValue(input[2], &y)) {
+    return JS::Result<mozilla::Ok>(JS::Error());
+  }
+  // 9. If buffer is not empty, it contains either 12 or 18 bits. If it contains 12 bits, then
+  // discard the last four and interpret the remaining eight as an 8-bit big-endian number. If it
+  // contains 18 bits, then discard the last two and interpret the remaining 16 as two 8-bit
+  // big-endian numbers. Append the one or two bytes with values equal to those one or two numbers
+  // to output, in the same order.
+  output += (uint8_t(w << 2 | x >> 4));
+  output += (uint8_t(x << 4 | y >> 2));
+  return mozilla::Ok();
+}
+
+inline JS::Result<mozilla::Ok> base64Decode2to1(std::string_view input, std::string &output) {
+  uint8_t w, x;
+  // 8.1 Find the code point pointed to by position in the second column of Table 1: The Base 64
+  // Alphabet of RFC 4648. Let n be the number given in the first cell of the same row. [RFC4648]
+  if (!base64CharacterToValue(input[0], &w) || !base64CharacterToValue(input[1], &x)) {
+    return JS::Result<mozilla::Ok>(JS::Error());
+  }
+  // 9. If buffer is not empty, it contains either 12 or 18 bits. If it contains 12 bits, then
+  // discard the last four and interpret the remaining eight as an 8-bit big-endian number. If it
+  // contains 18 bits, then discard the last two and interpret the remaining 16 as two 8-bit
+  // big-endian numbers. Append the one or two bytes with values equal to those one or two numbers
+  // to output, in the same order.
+  output += (uint8_t(w << 2 | x >> 4));
+  return mozilla::Ok();
+}
+
+bool isAsciiWhitespace(char c) {
+  switch (c) {
+  case '\t':
+  case '\n':
+  case '\f':
+  case '\r':
+  case ' ':
+    return true;
+  default:
+    return false;
+  }
+}
+
+// https://infra.spec.whatwg.org/#forgiving-base64-decode
+JS::Result<std::string> forgivingBase64Decode(std::string_view data) {
+  // 1. Remove all ASCII whitespace from data.
+  // ASCII whitespace is U+0009 TAB, U+000A LF, U+000C FF, U+000D CR, or U+0020 SPACE.
+  auto hasWhitespace = std::find_if(data.begin(), data.end(), &isAsciiWhitespace);
+  std::string dataWithoutAsciiWhitespace;
+
+  if (hasWhitespace) {
```

Comment:
`hasWhitespace` will be an iterator, so you'll want to compare explicitly against `data.end()` to check if there's whitespace.
```suggestion
  if (hasWhitespace != data.end()) {
```

Reply:
I somehow lost that as part of a refactor - I previously had it as part of the definition of `hasWhitespace` itself so that it was in fact a boolean. Seeing your other suggestion on line 3805 where we can use the position of the first found whitespace to skip iterating the elements prior to it is a great reason to keep the iteration position around

Reply:
I applied this suggestion as part of [`ae81dca` (#227)](https://github.com/fastly/js-compute-runtime/pull/227/commits/ae81dca453df75f519a3af68046348497954ab0e)